File: adapters/paypay/ts/test/vault.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { createHmac } from "node:crypto";
import { verifyResponseToken } from "../src/vault.ts";

// Signs an HS256 JWT the way PayPay signs responseTokens (key = the
// Base64-DECODED API secret).
const KEY = Buffer.from("0123456789abcdef0123456789abcdef");
const API_SECRET = KEY.toString("base64");

function signJWT(key: Buffer, header: object, claims: object): string {
  const b64url = (o: object) => Buffer.from(JSON.stringify(o)).toString("base64url");
  const signing = `${b64url(header)}.${b64url(claims)}`;
  const sig = createHmac("sha256", key).update(signing).digest("base64url");
  return `${signing}.${sig}`;
}

const HS256 = { alg: "HS256", typ: "JWT" };
const claims = (overrides: Record<string, unknown> = {}) => ({
  result: "succeeded",
  userAuthorizationId: "ua_9",
  aud: "client_1",
  iss: "paypay.ne.jp",
  exp: 1700003600,
  ...overrides,
});
const OPTS = { apiKey: "client_1", apiSecret: API_SECRET, now: () => 1700000000 };

test("valid token yields the userAuthorizationId", async () => {
  assert.equal(await verifyResponseToken(signJWT(KEY, HS256, claims()), OPTS), "ua_9");
});

const valid = signJWT(KEY, HS256, claims());
const forged = Buffer.from(JSON.stringify(claims({ userAuthorizationId: "ua_ATTACKER" }))).toString("base64url");
const [h, , s] = valid.split(".");

const noExp = (() => {
  const { exp: _drop, ...rest } = claims();
  return signJWT(KEY, HS256, rest);
})();

const rejects: Record<string, string> = {
  "missing exp": noExp,
  "undecodable signature segment": valid.split(".").slice(0, 2).join(".") + ".!!!not-base64url!!!",
  "tampered claims": `${h}.${forged}.${s}`,
  "alg none": signJWT(KEY, { alg: "none" }, claims()),
  "wrong signing key": signJWT(Buffer.from("wrong-key"), HS256, claims()),
  expired: signJWT(KEY, HS256, claims({ exp: 1600000000 })),
  "audience mismatch": signJWT(KEY, HS256, claims({ aud: "someone_else" })),
  "wrong issuer": signJWT(KEY, HS256, claims({ iss: "evil.example" })),
  "user declined": signJWT(KEY, HS256, claims({ result: "declined" })),
  "not a JWT": "not-a-jwt",
};

for (const [name, token] of Object.entries(rejects)) {
  test(`rejects: ${name}`, async () => {
    await assert.rejects(verifyResponseToken(token, OPTS));
  });
}

File: adapters/paypay/vault_test.go
package paypay

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"strings"
	"testing"
	"time"
)

// signJWT builds an HS256 JWT the way PayPay signs responseTokens (key = the
// Base64-DECODED API secret).
func signJWT(t *testing.T, key []byte, header, claims map[string]any) string {
	t.Helper()
	h, _ := json.Marshal(header)
	c, _ := json.Marshal(claims)
	signing := base64.RawURLEncoding.EncodeToString(h) + "." + base64.RawURLEncoding.EncodeToString(c)
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(signing))
	return signing + "." + base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

func TestVault(t *testing.T) {
	key := []byte("0123456789abcdef0123456789abcdef")
	apiSecret := base64.StdEncoding.EncodeToString(key)
	newAdapter := func() *Adapter {
		a := New(WithCredentials("client_1", apiSecret, "m1"))
		a.now = func() time.Time { return time.Unix(1700000000, 0) }
		return a
	}
	hs256 := map[string]any{"alg": "HS256", "typ": "JWT"}
	claims := func(overrides map[string]any) map[string]any {
		c := map[string]any{
			"result": "succeeded", "userAuthorizationId": "ua_9",
			"aud": "client_1", "iss": "paypay.ne.jp", "exp": 1700003600,
		}
		for k, v := range overrides {
			c[k] = v
		}
		return c
	}

	t.Run("valid token yields the userAuthorizationId", func(t *testing.T) {
		got, err := newAdapter().Vault(signJWT(t, key, hs256, claims(nil)))
		if err != nil || got != "ua_9" {
			t.Errorf("got %q, %v", got, err)
		}
	})

	// Tamper: swap the claims segment after signing — the MAC must not verify.
	valid := signJWT(t, key, hs256, claims(nil))
	forged, _ := json.Marshal(claims(map[string]any{"userAuthorizationId": "ua_ATTACKER"}))
	parts := strings.Split(valid, ".")
	tampered := parts[0] + "." + base64.RawURLEncoding.EncodeToString(forged) + "." + parts[2]

	rejects := map[string]string{
		"tampered claims":   tampered,
		"alg none":          signJWT(t, key, map[string]any{"alg": "none"}, claims(nil)),
		"wrong signing key": signJWT(t, []byte("wrong-key"), hs256, claims(nil)),
		"expired":           signJWT(t, key, hs256, claims(map[string]any{"exp": 1600000000})),
		"audience mismatch": signJWT(t, key, hs256, claims(map[string]any{"aud": "someone_else"})),
		"wrong issuer":      signJWT(t, key, hs256, claims(map[string]any{"iss": "evil.example"})),
		"user declined":     signJWT(t, key, hs256, claims(map[string]any{"result": "declined"})),
		"not a JWT":         "not-a-jwt",
	}
	for name, token := range rejects {
		t.Run(name, func(t *testing.T) {
			if got, err := newAdapter().Vault(token); err == nil {
				t.Errorf("want error, got token %q", got)
			}
		})
	}

	t.Run("non-string payment method", func(t *testing.T) {
		if _, err := newAdapter().Vault(42); err == nil {
			t.Error("want error for non-string")
		}
	})
}
